engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -575,10 +584,22 @@ protected void serviceInit(Configuration conf) throws Exception
       this.appDataPushAgent = new AppDataPushAgent(dnmgr, appContext);
       addService(this.appDataPushAgent);
     }
-    // initialize all services added above
+    initApexPluginDispatcher();
+
+    // Initialize all services added above
     super.serviceInit(conf);
   }
 
+  public static final String PLUGINS_CONF_KEY = "apex.plugin.stram.plugins";
+  private void initApexPluginDispatcher()
+  {
+    PluginLocator<ApexPlugin> locator = new ChainedPluginLocator<>(new ServiceLoaderBasedPluginLocator<>(ApexPlugin.class),
+        new PropertyBasedPluginLocator<>(ApexPlugin.class, PLUGINS_CONF_KEY));
+    apexPluginDispatcher = new PerPluginExecutorApexPluginManager(locator, appContext, dnmgr, stats);
+    dnmgr.apexPluginDispatcher = apexPluginDispatcher;
+    addService(apexPluginDispatcher);
```

Comment:
Looks like there is no clear separation between manager and dispatcher currently. There is only a single entity in the implementation being used for registering plugins as well as dispatching events. How about combining the interfaces. If future if the flexibility is needed this can be two separate entities.